### PR TITLE
feat: implemented biometrics authentication for SecureCredentialsManager using androidx.biometrics package

### PR DESCRIPTION
**Pull request by desusai7**
### Migrated to use [BioMetricManager](https://developer.android.com/reference/androidx/biometric/BiometricManager) for performing Authentication before accessing credentials via SecureCredentialsManager

Breaking Changes: 

* Removed the `requireAuthentication` API which used to be a pre-requisite previously for performing authentication while retrieving Credentials as it is no longer needed.
* Constructor of Auth0 class is made private, and now you will have to use Auth0.getInstance(clientId, domain) to get an instance of Auth0, it would check if the instance with the given configuration exists already, if yes returns it, if not would create one and returns it.
* Added multiple overloads of getCredentials() & awaitCredentials() to the BaseCredentialsManager interface, now all the implementations of this interface would have to implemented these new methods using override as previously they used to define these methods directly in their implementations.
* await function of Request interface has been made abstract now and all the implementations of the interface should implement the method.
* Credentials class has been converted to a data class now this class can longer be extended and also the currentTimeInMillis property of Credentials class has been removed.

Other changes: 

* We've added a dependency on `androidx.biometrics` package to use the `BiometricsManager` for authentication purposes.
* Updated the constructor of `SecureCredentialsManager` to accept `fragmentActivity` & `authenticationOptions` which are utilised while performing authentication using `BiometricPrompt` before retrieving credentials.  It's no longer needed to call `requireAuthentication` before retrieving credentials.
* Added class [LocalAuthenticationManager](https://github.com/auth0/Auth0.Android/blob/feat/biometric/auth0/src/main/java/com/auth0/android/authentication/storage/LocalAuthenticationManager.kt) which consolidates all the logic for interacting with the `BiometricsManager`
* Added class [LocalAuthenticationOptions](https://github.com/auth0/Auth0.Android/blob/feat/biometric/auth0/src/main/java/com/auth0/android/authentication/storage/LocalAuthenticationOptions.kt) which helps to configure the behaviour of the `BiometricsManager` like `title`, `description` displayed within in the biometric prompt and the level of authentication required.
* Updated `CredentialsManagerException` to contain enum `Code` describing the reason for exception and allowing users to check all the cases of `CredentialsManagerException` and handle accordingly

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. Since this library has unit testing, tests should be added for new functionality and existing tests should complete without errors.

- [x] This change adds unit test coverage
- [ ] This change adds integration test coverage
- [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors

### CODE REVIEW

File: auth0/src/main/java/com/auth0/android/authentication/storage/BaseCredentialsManager.kt [hunk 1]
```diff
@@ -90,4 +85,30 @@ public abstract class BaseCredentialsManager internal constructor(
...
+}
+
+public abstract class SecuredCredentialsManager internal constructor(
+    authenticationClient: AuthenticationAPIClient,
+    storage: Storage,
+    jwtDecoder: JWTDecoder
+) : BaseCredentialsManager(
+    authenticationClient, storage, jwtDecoder
+) {
+    public abstract fun getCredentials(fragmentActivity: FragmentActivity, authenticationOptions: LocalAuthenticationOptions, callback: Callback<Credentials, CredentialsManagerException>)
```
Comment (poovamraj) [inline auth0/src/main/java/com/auth0/android/authentication/storage/BaseCredentialsManager.kt:112]: If FragmentActivity is required in both method, can't we have this as a class level dependency, that way you can have same abstract method for both cred managers.

We just have to ensure there is no context leak by having fragment activity at a class level

File: auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManagerException.kt [hunk 2]
```diff
@@ -6,10 +6,182 @@ import com.auth0.android.result.Credentials
...
  * Represents an error raised by the [CredentialsManager].
  */
-public class CredentialsManagerException internal constructor(
-    message: String,
-    cause: Throwable? = null
-) : Auth0Exception(message, cause) {
+public class CredentialsManagerException :
+    Auth0Exception {
+
+    internal enum class Code {
```
Comment (poovamraj) [inline auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManagerException.kt:12]: Nice cleanup 👍

**Review by poovamraj — COMMENTED** [review]
Please take not of all the breaking changes in each PR so that it can be collected together for the major release. We don't want to lose track of them

File: auth0/src/main/java/com/auth0/android/authentication/storage/BaseCredentialsManager.kt (hunk 1 above)
Comment (desusai7) [reply auth0/src/main/java/com/auth0/android/authentication/storage/BaseCredentialsManager.kt]: Agreed, updated the code accordingly, please take a look again.

**Comment by sgammon** [comment]
Hi, we are an auth0 customer building for Android, and we are interested in seeing this ship. Is there any update?

**Review by poovamraj — DISMISSED** [review]

**Review by poovamraj — APPROVED** [review]

**Comment by desusai7** [comment]
> Hi, we are an auth0 customer building for Android, and we are interested in seeing this ship. Is there any update?

@sgammon, we will be releasing it shortly.

**Comment by sgammon** [comment]
@desusai7 Awesome! Thank you so much, we're very excited! 😄

**Comment by desusai7** [comment]
@sgammon, we've released it as part of version `3.0.0-beta.0`.

**Comment by sgammon** [comment]
We are already using it! I haven't figured out how to use the new biometrics yet, but that is next on the list. I plan to check samples, test suites, etc., but if there is already documentation, please let me know 😄

**Comment by desusai7** [comment]
> We are already using it! I haven't figured out how to use the new biometrics yet, but that is next on the list. I plan to check samples, test suites, etc., but if there is already documentation, please let me know 😄

documentation exists for it [here](https://github.com/auth0/auth0.android/blob/main/EXAMPLES.md#requiring-authentication)

**Pull request merged**